### PR TITLE
Support TS4.9 `satisfies` operator

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

`babel-ts` parser only

Partial of #13516
Ref #13750

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
👎 For marking it's ready, you should check all places uses `TSAsExpression`, since they should apply same logic.

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -265,7 +271,12 @@ function needsParens(path, options) {
         case "AssignmentPattern":
           return (
             key === "left" &&
-            (node.type === "TSTypeAssertion" || node.type === "TSAsExpression")
+            (node.type === "TSTypeAssertion" ||
+              node.type === "TSAsExpression" ||
+              // babel-parser cannot parse `satisfies` operator in left side of assignment
+              // https://github.com/babel/babel/issues/15095
```
Comment (sosukesuzuki) [inline src/language-js/needs-parens.js:277]: https://github.com/babel/babel/issues/15095

File: src/language-js/needs-parens.js [hunk 2]
```diff
@@ -227,14 +228,19 @@ function needsParens(path, options) {
...
+    case "TSSatisfiesExpression":
     case "LogicalExpression":
       switch (parent.type) {
         case "TSAsExpression":
           // example: foo as unknown as Bar
           return node.type !== "TSAsExpression";
 
+        case "TSSatisfiesExpression":
+          // example: foo satisfies unknown satisfies Bar
+          return node.type !== "TSSatisfiesExpression";
```
Comment (fisker) [inline src/language-js/needs-parens.js:240]: Maybe mixed should not add parens either

**Prettier pr-13764**
[Playground link](https://deploy-preview-13764--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEAzCEAEBnAhjAS21QLm0wFcoBrKCAdyh3yJLMwCFcAnAHSnRZc5KrQZNhnHv0HNCxUiJp1GmSVz4CMapWNV55bchpAAaEBAAOhaNmSge3BgAUeCOylwAbergCeduYARty4YNRwMADKuAC2cAAyBFBwyKje2HDBoeGRUZZhyQDmyDDcFFkgcLFBcAAmdfUJuFBFFLhFcABiENyx+IStyCC4FDAQZiAAFjCxXgDqUwTw2AVgcFHuywQAbst+w2DYgSDJmdwwzqFF-WkZlQBW2AAeUcVecACKFBDwd16ZcwFbjnYZBXC1LwAWhgJ0s3GSMHmBDqMCmyAAHAAGIFOTLzUKWYbwshwbg7VLmACOP3gVysHhG2ChKXq9Um3DgNIInKuHVuSHSAMqmViBFK5RF7y+tNSgvu5hgEORqPRSAATIrQgQvMUAMIQWICkZeLyTCiZAAqEI8QsBIB2FQAklBGrAomAEdYAIKuqIwPwff6ZAC+IaAA)

```sh
--parser babel-ts
```

**Input:**

```tsx
foo satisfies unknown satisfies Bar
foo as unknown as Bar
foo satisfies unknown as Bar
foo as unknown satisfies Bar
```

**Output:**

```tsx
foo satisfies unknown satisfies Bar;
foo as unknown as Bar;
(foo satisfies unknown) as Bar;
(foo as unknown) satisfies Bar;

```

File: src/language-js/print/typescript.js [hunk 3]
```diff
@@ -515,6 +515,8 @@ function printTypescript(path, options, print) {
       return printJSDocType(path, print, /* token */ "!");
     case "TSInstantiationExpression":
       return [print("expression"), print("typeParameters")];
+    case "TSSatisfiesExpression":
+      return [print("expression"), " satisfies ", print("typeAnnotation")];
```
Comment (fisker) [inline src/language-js/print/typescript.js:519]: Merge this into `TSAsExpression` too? There are extra check there.

Comment (fisker) [reply src/language-js/print/typescript.js:519]: https://github.com/prettier/prettier/pull/10341

File: src/language-js/needs-parens.js [hunk 4]
```diff
@@ -227,14 +228,23 @@ function needsParens(path, options) {
...
+        case "TSSatisfiesExpression":
+          // examples:
+          //   foo as unknown as Bar
+          //   foo satisfies unknown satisfies Bar
+          //   foo satisfies unknown as Bar
+          //   foo as unknown satisfies Bar
+          return (
+            node.type !== "TSAsExpression" &&
+            node.type !== "TSSatisfiesExpression"
+          );
```
Comment (fisker) [inline src/language-js/needs-parens.js:244]:
```suggestion
          return !isTSTypeExpression(node);
```

File: src/language-js/needs-parens.js [hunk 5]
```diff
@@ -265,7 +275,12 @@ function needsParens(path, options) {
         case "AssignmentPattern":
           return (
             key === "left" &&
-            (node.type === "TSTypeAssertion" || node.type === "TSAsExpression")
+            (node.type === "TSTypeAssertion" ||
+              node.type === "TSAsExpression" ||
+              // babel-parser cannot parse `satisfies` operator in left side of assignment
+              // https://github.com/babel/babel/issues/15095
+              // TODO: Add tests after the bug is fixed
+              node.type === "TSSatisfiesExpression")
```
Comment (fisker) [inline src/language-js/needs-parens.js:283]: We can reuse `isTSTypeExpression` here too.

**Review by fisker — APPROVED** [review]
Looks very good, thanks for the hard work, we can either wait for babel or update later.

**Comment by fisker** [comment]
FYI: @thorn0 may not agree to use `isTSTypeExpression` https://github.com/babel/babel/pull/14211#issuecomment-1289348178

**Comment by kachkaev** [comment]
Would it be possible to release this in 2.x? Upgrading to Prettier 3.0 may be more involved, so another 2.x release will allow folks to support TS 4.9 much faster.

**Comment by karlhorky** [comment]
This is great, thanks @sosukesuzuki 🙌 Will this be released in `prettier@2.8.0`?

**Comment by sosukesuzuki** [comment]
Maybe yes.

**Comment by hiyangguo** [comment]
> This is great, thanks @sosukesuzuki 🙌 Will this be released in `prettier@2.8.0`?

When will it be released?

**Comment by kachkaev** [comment]
@hiyangguo please follow https://github.com/prettier/prettier/issues/13792 and https://github.com/prettier/prettier/pull/13813

**Pull request merged**